### PR TITLE
Fix ListObjects aborting after 3 minute on async request

### PR DESCRIPTION
**Pull request by klauspost**
## Description

When async listing, if the first request does not return within 3 minutes, it is stopped, since it isn't being kept alive.

Keep updating `lastHandout` while we are waiting for the initial request to be fulfilled.

## How to test this PR?

Insert huge delays (1 second+ per entry) in returning entries from storage.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: cmd/metacache-server-pool.go [hunk 1]
```diff
@@ -419,6 +419,26 @@ func (z *erasureServerPools) listAndSave(ctx context.Context, o *listPathOptions
 		o.debugln("listAndSave: listing", o.ID, "finished with ", err)
 	}(*o)
 
+	// Keep alive while initial request is running.
+	go func(ctx context.Context, rpc *peerRESTClient, meta metacache) {
+		// Continuously update while we wait.
+		t := time.NewTicker(metacacheMaxClientWait / 10)
+		defer t.Stop()
+		select {
```
Comment (marktheunissen) [inline cmd/metacache-server-pool.go]: @klauspost Is this missing a `for {` wrapping the select, to keep it running continuously?

Comment (harshavardhana) [reply cmd/metacache-server-pool.go]: yes

Comment (klauspost) [reply cmd/metacache-server-pool.go]: Well, since I copy+pasted....

Comment (klauspost) [reply cmd/metacache-server-pool.go]: Good spot! Moving the shared code to a function.

**Comment by klauspost** [comment]
Tested with `time.Sleep(1000 * time.Millisecond)` when returning entries. Listing is kept alive even if there are 499 (10m timeouts) from mc.

**Review by harshavardhana — APPROVED** [review]

File: cmd/metacache-server-pool.go [hunk 2]
```diff
@@ -449,5 +440,10 @@ func (z *erasureServerPools) listAndSave(ctx context.Context, o *listPathOptions
 		xioutil.SafeClose(saveCh)
 	}()
 
-	return filteredResults()
+	entries, err = filteredResults()
+	if err == nil && err != io.EOF {
```
Comment (vadmeste) [inline cmd/metacache-server-pool.go]: isn't this OR rather than AND ?

Comment (harshavardhana) [reply cmd/metacache-server-pool.go]:
```suggestion
	if err == nil || err != io.EOF {
```

Comment (klauspost) [reply cmd/metacache-server-pool.go]: Yeah, it is a bit superfluous. Probably it should be `|| err == io.EOF` just to double check.

Comment (klauspost) [reply cmd/metacache-server-pool.go]: Let's check it.. Adds some extra safety.

**Review by vadmeste — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**